### PR TITLE
feat(containerregistry): add new check `containerregistry_not_publicly_accessible `

### PR DESCRIPTION
**Pull request by johannes-engler-mw**
### Context

Unrestricted network access to Azure Container registry should be restricted to enhance the security.

### Description

This PR adds an test for restricted network access to an Azure Container Registry.

### Checklist

- Are there new checks included in this PR? Yes
    - If so, do we need to update permissions for the provider? No
- [X] Review if the code is being covered by tests.
- [X] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [x] Review if backport is needed.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.metadata.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  "Provider": "azure",
+  "CheckID": "containerregistry_network_access_restricted",
+  "CheckTitle": "Restrict public network access to the Container Registry",
+  "CheckType": [],
+  "ServiceName": "containerregistry",
+  "SubServiceName": "",
+  "ResourceIdTemplate": "",
+  "Severity": "high",
+  "ResourceType": "ContainerRegistry",
+  "Description": "Ensure that public network access to the Azure Container Registry is restricted to enhance security. This prevents unauthorized access and improves the overall security posture of the container registry.",
```
Comment (puchy22) [inline prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.metadata.json]:
```suggestion
  "Description": "Ensure that public network access to the Azure Container Registry is restricted.",
```

File: prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    containerregistry_client,
+)
+
+
+class containerregistry_network_access_restricted(Check):
+    def execute(self) -> list[Check_Report_Azure]:
+        findings = []
+
+        for subscription, registries in containerregistry_client.registries.items():
+            for registry_id, ContainerRegistryInfo in registries.items():
```
Comment (puchy22) [inline prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.py]:
```suggestion
            for registry_id, container_registry_info in registries.items():
```

Please, use lower case and underscores for variables to follow our code style.

File: prowler/providers/azure/services/containerregistry/containerregistry_service.py [hunk 3]
```diff
@@ -39,11 +39,14 @@ def _get_container_registries(self):
                                 admin_user_enabled=getattr(
                                     registry, "admin_user_enabled", False
                                 ),
+                                network_rule_set=getattr(
+                                    registry, "network_rule_set", ""
```
Comment (puchy22) [inline prowler/providers/azure/services/containerregistry/containerregistry_service.py]:
```suggestion
                                    registry, "network_rule_set", None
```

File: prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.metadata.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+  "Provider": "azure",
+  "CheckID": "containerregistry_network_access_restricted",
```
Comment (puchy22) [inline prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.metadata.json]:
```suggestion
  "CheckID": "containerregistry_not_publicly_accessible",
```

File: prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+from prowler.lib.check.models import Check, Check_Report_Azure
+from prowler.providers.azure.services.containerregistry.containerregistry_client import (
+    containerregistry_client,
+)
+
+
+class containerregistry_network_access_restricted(Check):
```
Comment (puchy22) [inline prowler/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted.py]:
```suggestion
class containerregistry_not_publicly_accessible(Check):
```

File: tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+from azure.mgmt.containerregistry import models
+
+from tests.providers.azure.azure_fixtures import (
+    AZURE_SUBSCRIPTION_ID,
+    set_mocked_azure_provider,
+)
+
+
+class TestContainerRegistryNetworkAcessRestricted:
```
Comment (puchy22) [inline tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py]: Tenst
```suggestion
class Test_containerregistry_not_publicly_accessible:
```

File: prowler/providers/azure/services/containerregistry/containerregistry_service.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 from dataclasses import dataclass
 
-from azure.mgmt.containerregistry import ContainerRegistryManagementClient
+from azure.mgmt.containerregistry import ContainerRegistryManagementClient, models
```
Comment (puchy22) [inline prowler/providers/azure/services/containerregistry/containerregistry_service.py]:
```suggestion
from azure.mgmt.containerregistry import ContainerRegistryManagementClient
from azure.mgmt.containerregistry import NetworkRuleSet
```

File: prowler/providers/azure/services/containerregistry/containerregistry_service.py [hunk 8]
```diff
@@ -85,4 +88,5 @@ class ContainerRegistryInfo:
     login_server: str
     public_network_access: str
     admin_user_enabled: bool
-    monitor_diagnostic_settings: list[DiagnosticSetting] = None
+    network_rule_set: models.NetworkRuleSet
```
Comment (puchy22) [inline prowler/providers/azure/services/containerregistry/containerregistry_service.py]:
```suggestion
    network_rule_set: NetworkRuleSet
```

File: tests/providers/azure/services/containerregistry/containerregistry_admin_user_disabled/containerregistry_admin_user_disabled_test.py [hunk 9]
```diff
@@ -57,6 +57,8 @@ def test_container_registry_admin_user_enabled(self):
                         login_server="mock_login_server.azurecr.io",
                         public_network_access="Enabled",
                         admin_user_enabled=True,
+                        network_rule_set="",
```
Comment (puchy22) [inline tests/providers/azure/services/containerregistry/containerregistry_admin_user_disabled/containerregistry_admin_user_disabled_test.py]:
```suggestion
                        network_rule_set=None,
```

File: tests/providers/azure/services/containerregistry/containerregistry_admin_user_disabled/containerregistry_admin_user_disabled_test.py [hunk 10]
```diff
@@ -111,11 +113,14 @@ def test_container_registry_admin_user_disabled(self):
                         login_server="mock_login_server.azurecr.io",
                         public_network_access="Enabled",
                         admin_user_enabled=False,
+                        network_rule_set="",
```
Comment (puchy22) [inline tests/providers/azure/services/containerregistry/containerregistry_admin_user_disabled/containerregistry_admin_user_disabled_test.py]:
```suggestion
                        network_rule_set=None,
```

File: tests/providers/azure/services/containerregistry/containerregistry_service_test.py [hunk 11]
```diff
@@ -34,6 +34,7 @@ def test_get_container_registry(self):
                         login_server="mock_login_server.azurecr.io",
                         public_network_access="Enabled",
                         admin_user_enabled=True,
+                        network_rule_set="",
```
Comment (puchy22) [inline tests/providers/azure/services/containerregistry/containerregistry_service_test.py]:
```suggestion
                        network_rule_set=None,
```

File: tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
+from unittest import mock
+from unittest.mock import MagicMock
+from uuid import uuid4
+
+from azure.mgmt.containerregistry import models
```
Comment (puchy22) [inline tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py]:
```suggestion
from azure.mgmt.containerregistry.models import NetworkRuleSet
```

File: tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+                    registry_id: ContainerRegistryInfo(
+                        id=registry_id,
+                        name="mock_registry",
+                        location="westeurope",
+                        resource_group="mock_resource_group",
+                        sku="Basic",
+                        login_server="mock_login_server.azurecr.io",
+                        public_network_access="Enabled",
+                        admin_user_enabled=True,
+                        network_rule_set=models.NetworkRuleSet(default_action="Allow"),
```
Comment (puchy22) [inline tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py]:
```suggestion
                        network_rule_set=NetworkRuleSet(default_action="Allow"),
```

File: tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+                    registry_id: ContainerRegistryInfo(
+                        id=registry_id,
+                        name="mock_registry",
+                        location="westeurope",
+                        resource_group="mock_resource_group",
+                        sku="Basic",
+                        login_server="mock_login_server.azurecr.io",
+                        public_network_access="Enabled",
+                        admin_user_enabled=False,
+                        network_rule_set=models.NetworkRuleSet(default_action="Deny"),
```
Comment (puchy22) [inline tests/providers/azure/services/containerregistry/containerregistry_network_access_restricted/containerregistry_network_access_restricted_test.py]:
```suggestion
                        network_rule_set=NetworkRuleSet(default_action="Deny"),
```

**Review by puchy22 — CHANGES_REQUESTED** [review]

**Comment by johannes-engler-mw** [comment]
@puchy22 I have implemented your suggestions, please check again.

**Review by puchy22 — DISMISSED** [review]
LGTM! 👍

**Review by MrCloudSec — APPROVED** [review]

**Pull request merged**